src/main/java/net/fabricmc/mappingio/format/FeatureSet.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) 2024 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.mappingio.format;
+
+public interface FeatureSet {
+	boolean hasNamespaces();
+	MetadataSupport fileMetadata();
+	MetadataSupport elementMetadata();
+	NameSupport packages();
+	NameSupport classes();
+	MemberSupport fields();
+	MemberSupport methods();
+	LocalSupport args();
+	LocalSupport vars();
+	ElementCommentSupport elementComments();
+	boolean hasFileComments();
+
+	default boolean supportsPackages() {
+		return packages().srcNames() != SupportLevel.UNSUPPORTED
+				|| packages().dstNames() != SupportLevel.UNSUPPORTED;
+	}
+
+	default boolean supportsClasses() {
+		return classes().srcNames() != SupportLevel.UNSUPPORTED
+				|| classes().dstNames() != SupportLevel.UNSUPPORTED;
+	}
+
+	default boolean supportsFields() {
+		return FeatureSetHelper.isSupported(fields());
+	}
+
+	default boolean supportsMethods() {
+		return FeatureSetHelper.isSupported(methods());
+	}
+
+	default boolean supportsArgs() {
+		return FeatureSetHelper.isSupported(args());
+	}
+
+	default boolean supportsVars() {
+		return FeatureSetHelper.isSupported(vars());
+	}
+
+	enum MetadataSupport {
+		/** No metadata at all. */
+		NONE,
+
+		/** Only some select properties.  */
+		FIXED,
+
+		/** Arbitrary metadata may be attached. */
+		ARBITRARY
+	}
+
+	enum SupportLevel {
+		REQUIRED,
+		OPTIONAL,
+		UNSUPPORTED
+	}
+
+	interface NameSupport {
+		SupportLevel srcNames();
+		SupportLevel dstNames();
+	}
+
+	interface DescSupport {
+		SupportLevel srcDescs();
+		SupportLevel dstDescs();
+	}
+
+	interface MemberSupport extends NameSupport, DescSupport {
+	}
+
+	interface LocalSupport extends NameSupport, DescSupport {
+		SupportLevel positions();
+		SupportLevel lvIndices();
+		SupportLevel lvtRowIndices();
+		SupportLevel startOpIndices();
+		SupportLevel endOpIndices();
```

Comment:
Just curious, where are these terms from? I have a few naming suggestions.
 - `lvIndice` should be `slots`, "slot" is what Java officially uses for its java.lang.invoke and java.lang.classfile APIs, where long/double count as 2 slots.
 - `lvtRowIndices` should just be `lvtIndices`: it's the index of this local variable in the [`LocalVariableTable`](https://docs.oracle.com/javase/specs/jvms/se22/html/jvms-4.html#jvms-4.7.13) and/or the [`LocalVariableTypeTable`](https://docs.oracle.com/javase/specs/jvms/se22/html/jvms-4.html#jvms-4.7.14) attribute, difference being LVT stores descriptors while LVTT stores signatures.
 - `*OpIndices` can become `*Bci`, short for "bytecode index", this is what Java's Class-File API uses for an offset into the `Code` byte array.

Reply:
They're derived from the existing MIO terminology (see [here](https://github.com/FabricMC/mapping-io/blob/c5fa1758966f65c03de911e8bddfa66b2498e92e/src/main/java/net/fabricmc/mappingio/MappingVisitor.java#L108-L125) and [here](https://github.com/FabricMC/mapping-io/blob/c5fa1758966f65c03de911e8bddfa66b2498e92e/src/main/java/net/fabricmc/mappingio/tree/MappingTreeView.java#L348-L351)). I'm not opposed to changing these terms, depends on what @sfPlayer1 thinks

Reply:
We just had a little [discussion on Discord](https://discord.com/channels/507304429255393322/1032620376276074497/1231546506734211203), and he isn't particularly keen on changing these names.

Personally, I think:
- `lvIndex`: Would indeed make more sense as `slot` or `slotIndex`. These directly convey that this number is not about some sort of local variable count, but the variable's slot.
- `lvtRowIndex`: The "row" part disambiguates it from the various `index`-named attributes of the LVT, and makes clear that we're referring to the index of the variable entry within the LVT itself - IMO the name's worth keeping.
- `startOpIndex` and `endOpIndex`: These are tied to ASM's specific implementation of counting op-codes, which is probably incompatible with the Java Classfile API's one. This needs further investigation, but if anything, BCIs would probably be an addition rather than a replacement for the existing ASM op-code indices.

So all in all valid critique, but not worth breaking the existing API over.